KomaMRIBase/src/datatypes/phantom/motion/ArbitraryMotion.jl

```diff
@@ -43,106 +60,87 @@ julia> motion = ArbitraryMotion(
-    dz::Array{T,2}
-    ux::Vector{LinearInterpolator{T,V}}
-    uy::Vector{LinearInterpolator{T,V}}
-    uz::Vector{LinearInterpolator{T,V}}
-end
-
-function ArbitraryMotion(
-    period_durations::AbstractVector{T},
-    dx::AbstractArray{T,2},
-    dy::AbstractArray{T,2},
-    dz::AbstractArray{T,2},
-) where {T<:Real}
-    @warn "Note that ArbitraryMotion is under development so it is not optimized so far" maxlog = 1
-    Ns = size(dx)[1]
-    num_pieces = size(dx)[2] + 1
-    limits = times(period_durations, num_pieces)
-
-    #! format: off
-    Δ = zeros(Ns,length(limits),4)
-    Δ[:,:,1] = hcat(repeat(hcat(zeros(Ns,1),dx),1,length(period_durations)),zeros(Ns,1))
-    Δ[:,:,2] = hcat(repeat(hcat(zeros(Ns,1),dy),1,length(period_durations)),zeros(Ns,1))
-    Δ[:,:,3] = hcat(repeat(hcat(zeros(Ns,1),dz),1,length(period_durations)),zeros(Ns,1))
-   
-    etpx = [extrapolate(interpolate((limits,), Δ[i,:,1], Gridded(Linear())), Periodic()) for i in 1:Ns]
-    etpy = [extrapolate(interpolate((limits,), Δ[i,:,2], Gridded(Linear())), Periodic()) for i in 1:Ns]
-    etpz = [extrapolate(interpolate((limits,), Δ[i,:,3], Gridded(Linear())), Periodic()) for i in 1:Ns]
-    #! format: on
-
-    return ArbitraryMotion(period_durations, dx, dy, dz, etpx, etpy, etpz)
+struct ArbitraryMotion{T} <: MotionModel{T}
+    t_start::T
+    t_end::T
+    dx::AbstractArray{T}
+    dy::AbstractArray{T}
+    dz::AbstractArray{T}
 end
 
 function Base.getindex(
     motion::ArbitraryMotion, p::Union{AbstractRange,AbstractVector,Colon}
 )
-    fields = []
-    for field in fieldnames(ArbitraryMotion)
-        if field in (:dx, :dy, :dz)
-            push!(fields, getfield(motion, field)[p, :])
-        elseif field in (:ux, :uy, :uz)
-            push!(fields, getfield(motion, field)[p])
-        else
-            push!(fields, getfield(motion, field))
-        end
-    end
-    return ArbitraryMotion(fields...)
+    return ArbitraryMotion(motion.t_start, motion.t_end, motion.dx[p,:], motion.dy[p,:], motion.dz[p,:])
 end
 
 Base.:(==)(m1::ArbitraryMotion, m2::ArbitraryMotion) = reduce(&, [getfield(m1, field) == getfield(m2, field) for field in fieldnames(ArbitraryMotion)])
 Base.:(≈)(m1::ArbitraryMotion, m2::ArbitraryMotion)  = reduce(&, [getfield(m1, field) ≈ getfield(m2, field) for field in fieldnames(ArbitraryMotion)])
 
 function Base.vcat(m1::ArbitraryMotion, m2::ArbitraryMotion)
     fields = []
-    @assert m1.period_durations == m2.period_durations "period_durations of both ArbitraryMotions must be the same"
-    for field in
-        Iterators.filter(x -> !(x == :period_durations), fieldnames(ArbitraryMotion))
+    @assert (m1.t_start == m2.t_start) && (m1.t_end == m2.t_end) "starting and ending times must be the same"
+    for field in (:dx, :dy, :dz)
         push!(fields, [getfield(m1, field); getfield(m2, field)])
     end
-    return ArbitraryMotion(m1.period_durations, fields...)
+    return ArbitraryMotion(m1.t_start, m1.t_end, fields...)
```

Comment:
Simplify.

Reply:
Resolved in [cfeddb2](https://github.com/JuliaHealth/KomaMRI.jl/pull/408/commits/cfeddb221565468eb6581ad987bde530010c0d39)